libpromises/evalfunction.c

```diff
@@ -1124,7 +1124,25 @@ static StringSet *VariablesMatching(const EvalContext *ctx, VariableTableIterato
 
             if (pass)
             {
-                StringSetAdd(matching, expr);
+                JsonElement *data = NULL;
+                bool allocated = false;
+                if (collect_full_data)
+                {
+                    data = VarRefValueToJson(ctx, fp, v->ref, NULL, 0, true, &allocated);
+                }
+
+                if (data == NULL)
+                {
+                    JsonObjectAppendString(matching, expr, "");
```

Comment:
When we don't collect the full variable data, we create a key-value map JsonObject with empty strings as the values. It will be destroyed soon afterwards, but the code is cleaner IMHO if we do it this way than if we make a JsonArray in one branch and a JsonObject in the other branch. The empty strings provide assurance that serializing this JsonObject (e.g. for logging) will not create problems. The extra memory usage from the empty strings is negligible.